### PR TITLE
Treeshap

### PR DESCRIPTION
**Pull request by rayeaster**
implementation for https://github.com/haifengl/smile/issues/515

### CODE REVIEW

File: core/src/test/java/smile/regression/TreeShapTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		}
+		return meanShap;
+	}
+
+	@Test
+	public void testBostonHousing() {
+		List<String> sortedFields = new ArrayList<String>();
+		double[] shaps = test(Loss.ls(), "Boston Housing", BostonHousing.formula, BostonHousing.data, sortedFields, 100, 0.01, 1);
+		assertTrue(sortedFields.get(0).equals("LSTAT"));
+		assertEquals(2.3119, shaps[shaps.length - 1], 1E-4);
```
Comment (haifengl) [inline core/src/test/java/smile/regression/TreeShapTest.java:88]: How do you get these numbers such as 2.3119?

Comment (rayeaster) [reply core/src/test/java/smile/regression/TreeShapTest.java:88]: run the test will show these numbers, also compared to the python result shows similar value (although not exactly same).

```
import xgboost
import shap
import numpy as np

# load JS visualization code to notebook
shap.initjs()
# train XGBoost model
X,y = shap.datasets.boston()
model = xgboost.train({"learning_rate": 0.01}, xgboost.DMatrix(X, label=y), 100)
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X)
shap_sum = np.sum(np.absolute(shap_values),axis=0)
shap_sum = shap_sum / y.size
print(shap_sum)            
shap.summary_plot(shap_values, X, plot_type="bar")
```
bostonShap

**Comment by haifengl** [comment]
You should not merge master. Now there are tons of file changed. It is my responsibility to merge feature branch.

File: core/src/test/java/smile/regression/TreeShapTest.java (hunk 1 above)
Comment (haifengl) [reply core/src/test/java/smile/regression/TreeShapTest.java:88]: Can you put the exact number from python code into a comment line? Thanks

**Comment by rayeaster** [comment]
> You should not merge master. Now there are tons of file changed. It is my responsibility to merge feature branch.

aha, fetch latest code from master and do a rebase on this PR. maybe because shap branch is not sync-up with master

**Comment by rayeaster** [comment]
python result for boston housing dataset:

```
CRIM      4.7090670e-01
ZN        1.1862384e-03
INDUS     3.6680367e-02
CHAS      0.0000000e+00
NOX       2.9865596e-01
RM        1.6684741e+00
AGE       3.4423586e-02
DIS       2.4549793e-01
RAD       1.1467161e-02
TAX       4.9007054e-02
PTRATIO   9.5596902e-02
B         8.1507929e-02
LSTAT     2.5720918e+00
```

**Pull request closed without merging**